biogeochem/FatesSoilBGCFluxMod.F90

```diff
@@ -0,0 +1,1114 @@
+          cpatch => cpatch%younger
+       end do
+       
+    end do
+    
+    ! We can exit if this is a c-only simulation
+    if(hlm_parteh_mode.ne.prt_cnp_flex_allom_hyp) then
+       ! These can now be zero'd
+       do s = 1, nsites
+          bc_in(s)%plant_n_uptake_flux(:,:) = 0._r8
+          bc_in(s)%plant_p_uptake_flux(:,:) = 0._r8
+       end do
+       return
+    end if
+
+    
+    do s = 1, nsites
+
+       ! If the plant's PFT has a non-zero positive prescribed rate,
+       ! use that instead of the fully coupled version.
+
+       if (n_uptake_mode.eq.prescribed_n_uptake) then
+          cpatch => sites(s)%oldest_patch
+          do while (associated(cpatch))
+             ccohort => cpatch%tallest
+             do while (associated(ccohort))
+                pft = ccohort%pft
+                ccohort%daily_n_uptake = &
+                     EDPftvarcon_inst%prescribed_nuptake(pft) * &
+                     GetPlantDemand(ccohort,nitrogen_element)
+                ccohort => ccohort%shorter
+             end do
+             cpatch => cpatch%younger
+          end do
+       end if
+       
+       if (p_uptake_mode.eq.prescribed_p_uptake) then
+          cpatch => sites(s)%oldest_patch
+          do while (associated(cpatch))
+             ccohort => cpatch%tallest
+             do while (associated(ccohort))
+                pft = ccohort%pft
+                ccohort%daily_p_uptake = &
+                     EDPftvarcon_inst%prescribed_puptake(pft) * &
+                     GetPlantDemand(ccohort,phosphorus_element)
+                ccohort => ccohort%shorter
+             end do
+             cpatch => cpatch%younger
+          end do
+       end if
+       
+       
+       ! If nutrient competition is sent to the BGC model as PFTs
+       ! and not as individual cohorts, we need to unravel the input
+       ! boundary condition and send to cohort.  We do this downscaling
+       ! by finding each cohort's fraction of total fine-root for the group
+       if(n_uptake_mode.eq.coupled_n_uptake .or. p_uptake_mode.eq.coupled_p_uptake)then
+
+          nlev_eff_soil = max(bc_in(s)%max_rooting_depth_index_col, 1)
+          
+          if(fates_ncomp_scaling.eq.pft_ncomp_scaling) then
+             
+             allocate(rootfrac_pft_decomp(numpft,bc_in(s)%nlevdecomp))
+             allocate(fnrt_c_pft(numpft))
+             
+             ! Construct the pft x decomp layer fine-root profile
+             ! USe the profile associated with hydraulic uptake! 
+             
+             do pft = 1, numpft
+                rootfrac_pft_decomp(pft,:) = 0._r8
+                call set_root_fraction(sites(s)%rootfrac_scr, pft, sites(s)%zi_soil)
+                
+                do j = 1,nlev_eff_soil
+                   id = bc_in(s)%decomp_id(j)  ! Map from soil layer to decomp layer    
+                   rootfrac_pft_decomp(pft,id) = rootfrac_pft_decomp(pft,id) + & 
+                        sites(s)%rootfrac_scr(j)
```

Comment:
how come rootfrac is a site level variable here?

Reply:
good catch @rosiealice.
The whole point of this array is to figure out, for each soil layer, the fraction of a cohort's root contribution to the sum within it's PFT.  That fraction dictates the share the cohort gets from the group's nutrient uptake.  However, the math is wrong.  Also, since all cohorts in a PFT have the same rooting profile (right now) calculating this fraction is not necessary.  Applying a fix.  (note that none of the tests have been using this scaling strategy.  Up to this point, I have been having all cohorts act as individual competitors in the soil model. This only applies to grouping them by PFT).

Reply:
This has now been updated.  For each layer, a cohort will get a share of the nutrient, proportional to its mass fraction of the total for the PFT.